File: client/types/payload/request/invitation.request.ts
import { z } from "zod";

export const InvitationRequest = z.object({
  username: z
    .string()
    .min(5, "Username is too short. Please enter at least 5 characters")
    .max(50, "Username is too long. Please enter no more than 50 characters")
    .nonempty("Username cannot be blank"),

  email: z
    .string()
    .email("Please provide a valid email address")
    .min(5, "Email is too short. Please enter at least 5 characters")
    .max(100, "Email is too long. Please enter no more than 100 characters")
    .nonempty("Email cannot be blank"),

  roles: z
    .array(
      z
        .string()
        .regex(
          /^ROLE_[A-Z_]+$/,
          "Each role must start with 'ROLE_' and contain only uppercase letters",
        ),
    )
    .min(1, "At least one role must be provided"),
});

export type InvitationRequest = z.infer<typeof InvitationRequest>;

File: backend/src/main/java/com/lerneon/backend/models/payload/request/InvitationRequest.java
package com.lerneon.backend.models.payload.request;

import jakarta.validation.constraints.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.List;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class InvitationRequest {
    @NotBlank(message = "Username cannot be blank")
    @Size(min = 5, message = "Username is too short. Please enter at least 5 characters")
    @Size(max = 50, message = "Username is too long. Please enter no more than 50 characters")
    private String username;

    @NotBlank(message = "Email cannot be blank")
    @Email(message = "Please provide a valid email address")
    @Size(min = 5, message = "Email is too short. Please enter at least 5 characters")
    @Size(max = 100, message = "Email is too long. Please enter no more than 100 characters")
    private String email;

    @NotNull(message = "Roles cannot be null")
    @NotEmpty(message = "At least one role must be provided")
    private List<@Pattern(
            regexp = "^ROLE_[A-Z_]+$",
            message = "Each role must start with 'ROLE_' and contain only uppercase letters"
    ) String> roles;
}
